### DISCUSSION TITLE
Prolog 5기 BE 도메인 분석

### DISCUSSION THREAD

**Discussion by nuyh99** [woowacourse/prolog#1275] [backend]
## 작성 방법
* 패키지 단위로 쓰레드를 생성한다.
* 해당 패키지에 대한 추가 내용이 있다면, 해당 쓰레드를 업데이트한다.

## 쓰레드 예시
```markdown
# {패키지명}
## {서비스명}
{정리 내용}
## {서비스명}
{정리 내용}
```

**Comment by nuyh99** [comment]
# Member
## MemberController
#### findMemberInfo
	패스변수로 전달되는 username을 통해 멤버를 조회하여 전달한다.
#### updateMember
	LoginMemeber 의 id와 인증 상태, 멤버의 이름, 업데이트할 정보를 통해 멤버를 업데이트한다.
	인증 상태가 익명일 경우 오류를 던진다.
	id로 조회한 멤버의 이름이 username이 동일하지 않을 경우 오류를 던진다.
	id로 조회한 멤버의 해당 멤버의 nickName, imageUrl을 업데이트한다
#### show ( admin only )
	id를 기준으로 내림차순으로 정렬하며 기본값으로 size는 10, page는 0으로 설정된 pageable의 정보를 통해 멤버를 조회하여 반환한다.
	@MemberOnly 는 aop를 사용해 멤버의 인증 정보를 검증하는데 사용됐었다.

## MemberRecationController
#### showScrap
        @MemberOnly 때문에 인증 정보가 익명이 아닌지 AOP 에서 검증
	멤버의 id와 pageable의 정보를 통해 스크랩한 학습로그들의 정보를 조회하여 반환한다.
#### registerScrap
        @MemberOnly 때문에 인증 정보가 익명이 아닌지 AOP 에서 검증
	멤버가 해당 학습로그를 스크랩했는지 확인하여 했을 경우 오류를 던진다.
	학습로그 ID로 학습로그를 조회한다. ( 없을 경우 오류를 던진다.)
	멤버 ID로 멤버를 조회한다. ( 없을 경우 오류를 던진다.)
	학습로그 스크랩을 저장하고 해당 정보를 반환한다.
#### unregisterScrap
        @MemberOnly 때문에 인증 정보가 익명이 아닌지 AOP 에서 검증
	스크랩된 학습로그를 조회하여 존재하지 않을 경우 오류를 던진다.
	해당 스크랩된 학습로그를 삭제한다.

**Comment by amaran-th** [comment]
# StudyLog

## ✅StudylogSessionService
### 👉syncStudylogSession()
옛날엔 Session 없이 Mission만 있었고, sturylog는 필드로 Mission을 **필수적으로** 가지고 있었음. 나중에 Session을 필드로 추가하게 되면서, Mission이 가진 Session 정보를 가지고 와서 studylog의 Session을 자동으로 셋팅하기 위해 구현됨.
(실제 서비스에서는 사용되지 않는 메서드)

## ✅ViewedStudyLogCookieGeneratorTest
### 👉generateCookie()

쿠키 헤더에 viewd 값을 추가하여 반환하는 메서드

### 👉isViewed()

입력받은 String studyLogIds가 "/studyId/"의 형식인지 여부를 반환하는 메서드

### 👉setViewedStudyLogCookie()

입력받은 String studyLogIds가 "/studyId/"의 형식이라면(isViewed()==true), 해당 정보를 기반으로 쿠키를 생성하고 response 객체의 header에 추가하는 메서드

## ✅StudylogService
### 👉findStudylogById

id 기반으로 `studylogRepository`에서 studylog 를 가져온다.

해당 studylog 가 없을 경우 `StudylogNotFoundException`가 발생한다.

### 👉retrieveStudylogById

Studylog 를 검색하는 전체 플로우를 담당하는 메서드
익명 또는 다른 유저가 글을 읽는다면 조회수가 끝없이 오르는 이유이다.

1. findStudylogById 를 통해서 Studylog를 가져온다.
2. 조회수를 증가시킨다.  
- 쿠키를 기반으로 해당 글을 과거에 읽었는지 확인한다. 읽었다면 true, 아니라면 false가 할당된다.
    - 해당 부분에서 쿠키가 제대로 작동하고 있지 않기 떄문에 항상 false 가 입력된다.
- 쿠키 기반 읽음 유무가 false 라면 조회수 증가 로직을 타게된다. 현재 이 로직을 항상 타게된다.
    - 익명유저라면 조회수를 증가시키고 종료한다. 
    - 로그인 유저라면 해당 글에 대한 주인인지 확인한다.
        - 만약 자신의 글이라면 조회수는 증가하지 않는다.
        - 다른 사람의 글이라면 조회수는 증가한다. 
3. 익명유저라면 종료한다.
4. 로그인 유저라면 만약 읽은 글의 기록을 저장하는 repository에 해당 studylog를 읽었다 저장한다.

### 👉deleteStudylog

유저의 아이디와 스터디로그의 아이디를 받아온다.

1. 유저와 스터디 로그가 없으면 예외가 터진다.
2. 유저의 스터디 로그가 아니면 예외가 터진다.

1. 태그서비스에서 스터디로그와 로그인 유저를 기반으로 Tags 를 가져온다.
2. 스터디 로그를 제거한다.
- 스터디 로그를 제거하면서 해당 스터디 로그를 읽었다는 StudylogRead도 함꼐 제거한다.
- 마찬가지로 해당 스터디로그에 대한 scrap 도 제거한다.
3. 3번 단계에서 가져왔던 Tags를 제거한다.
4. 스터디 로그 제거 이벤트를 발행한다.

### 👉updateStudyMission

1. 스터디 로그를 가져오고 없으면 예외발생
2. 로그인 맴버가 새당 스터디로그의 주인인지 확인
3. 아이디 기반으로 mission Service에서 미션을 가져온다
4. 1.번에서 가져온 스터디 로그에 3번에서 가져온 mission 으로 업데이트한다.

### 👉findStudylogs

스터디 로그에 대한 검색을 진행한다.

1. 만약 검색 요청에 studylog 에 대한 id 들이 있었다면 해당 studylog들을 반환하고 메서드는 종료된다.
2. 만약 id가 없고 키워드가 없다면 키워드를 제외한 정보를 바탕으로 studylog를 검색하고 반환한다.
- 해당 부분은 요청 정보에 표함된 정보를 `findStudylogsWithoutKeyword`에서 동적 쿼리를 생성하여 조건을 검색한다.
3. id가 없고 키워드가 있는 경우에는 `findBySearchKeyword`를 통해 StudylogDocument를 가져온다.
- StudylogDocument 는 Studylog 와 다르게 elastic search 를 기반으로 검색하는 studylog 를 뜻한다.
4. Es 로 찾아온 Document 속 아이디들을 바탕으로 StudylogRepository 를 통해 Studylog를 가져온다.

## ✅ StudylogTagService
### 👉findAll
저장된 모든 StudylogTag를 찾습니다.

###👉findTagsIncludedInStudylogs
studylog에서 사용되고 있는 tag들을 조회합니다.

## ✅TagService
### 👉findOrCreate
요청들에서 포함된 tag들을 찾고, 요청중에서 이미 디비에 저장되지 않은 Tag가 있다면
Tag를 생성해서, 디비에 저장하고 반환한다.

### 👉findTagsIncludedInStudylogs
studylog에서 사용되는 tag들을 조회하고, 중복을 제거 후 이름순으로 정렬해서 반환한다.

**Comment by hong-sile** [comment]


**Reply by nuyh99** [reply]
[아마란스의 Studylog](https://github.com/woowacourse/prolog/discussions/1275#discussioncomment-5901777)
여기에 업데이트 해야 하지 않나요...??!

**Reply by BGuga** [reply]
합쳐놨습니다

**Comment by java-saeng** [comment]
# Member, GroupMember(이하 GM), MemberGroup(이하 MG) 관계

Member : GM = 1 : N
MG : GM = 1: N

MG => 그룹, GM => 멤버와 그룹간의 맵핑 테이블

**Comment by wonyongChoi05** [comment]
https://wandering-niece-921.notion.site/1afb445baea0488e991495e68be59ee4

**Comment by java-saeng** [comment]
## Mission

![image](https://github.com/woowacourse/prolog/assets/62413589/bc76a45a-e7dd-4922-90f7-427d3be7e74c)

## Session

![image](https://github.com/woowacourse/prolog/assets/62413589/c45b71dd-302d-4d89-ad95-537e08b3483c)


### MissionService

#findAllWithMyMissionFirst
모든 미션과 나의 미션에 대한 차집합을 구하여, 미션 목록을 볼 때 나의 미션이 가장 위로 보여주기 위함


### SessionMemberService

SessionMember는 다대다 관계인 Member와 Session 을 풀어내기 위한 맵핑 테이블

### SessionService

#findAllWithMySessionFirst
모든 세션과 나의 세션에 대한 차집합을 구하여, 세션 목록을 볼 때 나의 세션이 가장 위로 보여주기 위함

**Comment by eunbii0213** [comment]
# Member

https://chocolate-innovation-7ff.notion.site/Member-e19a6361d23c4c37a0805df3dec00f1b

**Comment by nuyh99** [comment]
# Roadmap
## Quiz
<img width="1237" alt="image" src="https://github.com/woowacourse/prolog/assets/93072571/c40315f3-ced5-46a7-b0ba-0c80d6c4a04d">

- 로드맵에서 하위 키워드에 퀴즈가 존재한다.
- `create()` : 세션(백엔드 레벨 1) ID, 키워드(자료 구조) ID 두 가지를 받아서 퀴즈를 등록할 수 있다.
  - 퀴즈 등록은 admin 권한만 가능하다. (현재 검증 로직 없음)
- `findQuizById` : 퀴즈 ID로 퀴즈를 조회한다.
- `findQuizzesByKeyword` : 키워드(자료 구조) ID에 귀속된 퀴즈들을 조회한다.
- `updateQuiz` : 퀴즈 ID의 퀴즈를 수정한다.
- `deleteQuiz` : 퀴즈 ID의 퀴즈를 제거한다.

---

## NewSession
- 특정 커리큘럼에 대한 세션 정보를 관리할 때 사용한다.
- POST /curriculums/{curriculmId}/sessions, body = name (세션 이름)
	- 특정 커리큘럼에 대한 세션 정보를 추가한다.
	- 생성된 세션에 대한 정보를 반환한다. (Location = /sessions/{sessionId})
- GET /curriculums/{curriculmId}/sessions
	- 특정 커리큘럼에 대한 세션 정보를 반환한다, response = List<SessionsResponse> = (세션 아이디, 이름)
- PUT /curriculums/{curriculmId}/sessions/{sessionId}, body = name (변경할 세션 이름)
	- 특정 커리큘럼의 특정 세션에 대한 정보를 수정한다. 
	- 특정 세션이 존재하지 않으면 예외가 발생한다.
- DELETE /curriculums/{curriculmId}/sessions/{sessionId}
	- 특정 커리큘럼에 대한 특정 세션 정보를 제거한다. 
	- 특정 세션이 존재하지 않으면 예외가 발생한다.

### 리팩터링 필요한 부분
- Controller에서 사용하지 않는 파라미터가 존재함 (updateSession, deleteSession)
- 세션이 존재하는지 판단하는 부분에 대한 메서드 통일 필요 (find로 찾아올지, exist로 확인할지)

---

###Curriculum

Curriculum 에 대한 기본적 Crud 서비스
update, delete 시 Curriculum 가 존재하지 않을 경우 예외가 발생하는 것 외에는 특이사항 없음.

**Comment by splitCoding** [comment]
# update

## MemberTagApplicationListener

### onApplicationEvent(ContextRefreshedEvent event)

1. ContextRefreshedEvent 가 발생시 메서드가 실행된다.
    - ContextRefreshedEvent : ApplicationContext를 초기화하거나 새로 고칠 때 Spring은 ContextRefreshedEvent를 발생시킵니다.
2. UpdatedContents 테이블에서 “*MEMBER_TAG_UPDATE”* 를 content로 가지고 있는 로우를 조회한다.
    - 사실상 content로 “*MEMBER_TAG_UPDATE”* 밖에 가지지 못한다.
3. 해당 로우가 존재하고 updated 컬럼 값이 1일 경우 ( 업데이트 됨 ) 이미 업데이트가 진행되었기에 early return
4. 모든 스터디 로그의 멤버와 태그를 통하여 모든 멤버태그를 저장한다.
5. content로 “*MEMBER_TAG_UPDATE”*를 가지고 *updated*로 1을 가진 UpdatedContents를 저장한다.

**Comment by eunbii0213** [comment]
# Badge

https://chocolate-innovation-7ff.notion.site/Badge-0971882182cb4088a1bb5ee6c9435fbd

**Reply by java-saeng** [reply]
@eunbii0213  주디 이 부분 볼 수 있는 권한이 없는 것 같아요~

**Reply by eunbii0213** [reply]
@java-saeng  허걱~! 수정했습니다! 감사합니다!

**Comment by amaran-th** [comment]
# GithubLogService
Spring(Java)에서 Github 소셜 로그인을 구현하기 위해 알아둬야 할 인증 흐름은 다음과 같다. 

1. `https://github.com/login/oauth/authorize?client_id={발급받은 client_ID}` 주소를 사용자에게 띄워준다.(프론트에서 리다이렉션)
2. 사용자는 Github 로그인을 통해서 인증을 한다.
3. 인증에 성공하면, Github는 {우리가 설정한 콜백 URL}?code={인증코드} 로 **code값**을 쿼리 파라미터 형태로 보내준다.
4. 받은 code를 이용해서 `https://github.com/login/oauth/access_token` 로 {client_ID,client_Secret,code}를 POST요청으로 전송한다.
5. Github는 `access_token`을 응답해서 서버측으로 보내준다.
6. 서버는 `access_token`을 `https://api.github.com/user` 로 담아서 GET 요청을 보낸다.
7. Github는 로그인한 사용자의 정보를 서버측으로 보내준다.
8. 받은 사용자 정보를 사용한다.(회원가입 혹은 로그인)

**4번 단계부터가 서버에서 처리하는 동작이다.**

현재(2023-05-26) 기준 GithubLogService 클래스에 작성되어 있는 createToken() 메서드는 다음과 같다.

```java
private final GithubClient githubClient;
...
public TokenResponse createToken(TokenRequest tokenRequest) {
		String githubAccessToken = githubClient.getAccessTokenFromGithub(tokenRequest.getCode());
    GithubProfileResponse githubProfile = githubClient
        .getGithubProfileFromGithub(githubAccessToken);
    Member member = memberService.findOrCreateMember(githubProfile);
    String accessToken = jwtTokenProvider.createToken(member);
    return TokenResponse.of(accessToken);
}
```

GithubClient  클래스에 구현되어 있는 `getAccessTokenFromGithub()`는 Github Auth API를 호출하여 Github 로그인 인증 토큰을 받아온다.

 `getGithubPorfileFromGithub()`는 앞 라인에서 받아온 Github 로그인 인증 토큰으로 사용자의 Github 계정의 프로필 정보를 받아온다.(닉네임, 프로필 이미지 등)

이제 각 메서드의 상세 동작 메커니즘을 살펴보자.

### 4. 받은 code를 이용해서 `https://github.com/login/oauth/access_token` 로 {client_ID,client_Secret,code}를 POST요청으로 전송한다.

`getAccessTokenFromGithub()`

```java
public String getAccessTokenFromGithub(String code) {
    GithubAccessTokenRequest githubAccessTokenRequest = new GithubAccessTokenRequest(
        code,
        clientId,
        clientSecret
    );

    HttpHeaders headers = new HttpHeaders();
    headers.add("Accept", MediaType.APPLICATION_JSON_VALUE);

    HttpEntity<MultiValueMap<String, String>> httpEntity = new HttpEntity(
        githubAccessTokenRequest, headers);
    RestTemplate restTemplate = new RestTemplate();

    String accessToken = restTemplate
        .exchange(tokenUrl, HttpMethod.POST, httpEntity, GithubAccessTokenResponse.class)
        .getBody()
        .getAccessToken();
    if (accessToken == null) {
        throw new GithubApiFailException();
    }
    return accessToken;
}
```

Github Auth API를 호출하기 위해 `{client_ID,client_Secret,code}`를 데이터로 담은 요청 객체(HttpEntity)를 생성한다.

<aside>
💡 RestTemplate이란?
Spring에서 HTTP 통신을 RESTful 형식에 맞게 간편하게 사용할 수 있게 해주는 템플릿이다.
Rest API 서비스에 요청을 보내고 응답을 받을 수 있다.

- `exchange()` : 원하는 HTTP 메소드를 요청하고 결과를 ResponseEntity로 반환한다.
</aside>

```java
public <T> ResponseEntity<T> exchange(String url, HttpMethod method,
			@Nullable HttpEntity<?> requestEntity, Class<T> responseType, Object... uriVariables)
			throws RestClientException {
```

exchange() 메서드의 파라미터는 순서대로 요청을 보낼 **url**, HTTP **메서드**, **요청 객체**, Response data를 매핑할 **객체(dto)의 클래스**이다.

Github 로그인 인증 API를 호출하기 위해 `tokenUrl`에는 다음 url이 담겨있다.

```
https://github.com/login/oauth/access_token
```

해당 url에 `{client_ID,client_Secret,code}`을 담은 POST 요청을 보낸다.

### 5. Github는 `access_token`을 응답해서 서버측으로 보내준다.

이렇게 해서 응답받은 데이터에는 인증된 Github 계정의 액세스 토큰이 포함되어있다. 이를 추출한다.

```java
String githubAccessToken = githubClient.getAccessTokenFromGithub(tokenRequest.getCode());
```

### 6. 서버는 `access_token`을 `https://api.github.com/user` 로 담아서 GET 요청을 보낸다.

`getGithubProfileFromGithub()`

```java
public GithubProfileResponse getGithubProfileFromGithub(String accessToken) {
    HttpHeaders headers = new HttpHeaders();
    headers.add("Authorization", "token " + accessToken);

    HttpEntity httpEntity = new HttpEntity<>(headers);
    RestTemplate restTemplate = new RestTemplate();

    try {
        return restTemplate
            .exchange(profileUrl, HttpMethod.GET, httpEntity, GithubProfileResponse.class)
            .getBody();
    } catch (HttpClientErrorException e) {
        throw new GithubConnectionException();
    }
}
```

마찬가지로 profileUrl에는 다음 url 정보가 선언되어 있다.

```
https://api.github.com/user
```

위 url에 GET 요청을 보낸다.

### 7. Github는 로그인한 사용자의 정보를 서버 측으로 보내준다.

이렇게 해서 응답 받은 데이터에는 다음과 같은 회원 정보가 들어가 있다.

```java
public class GithubProfileResponse {

    @JsonProperty("name")
    private String nickname;
    @JsonProperty("login")
    private String loginName;
    @JsonProperty("id")
    private String githubId;
    @JsonProperty("avatar_url")
    private String imageUrl;
		
		...
```

### 8. 받은 사용자 정보를 사용한다.(회원 가입 혹은 로그인)

### 참고 게시글

[· Spring Boot Github 소셜 로그인 구현하기 (RestTemplate · WebClient)](https://inkyu-yoon.github.io/docs/Language/SpringBoot/GithubLogin)

---

```java
private final GithubClient githubClient;
...
public TokenResponse createToken(TokenRequest tokenRequest) {
		String githubAccessToken = githubClient.getAccessTokenFromGithub(tokenRequest.getCode());
    GithubProfileResponse githubProfile = githubClient
        .getGithubProfileFromGithub(githubAccessToken);
    Member member = memberService.findOrCreateMember(githubProfile);
    String accessToken = jwtTokenProvider.createToken(member);
    return TokenResponse.of(accessToken);
}
```
- memberService findOrCreateMember(GithubProfileResponse githubProfile) : githubId가 MemberRepo에 존재하면 조회해오고 없다면은 새로 만들어서 반환한다.
이는 편하게 로그인 하기 위함이다.
- jwtTokenProvider createToken(Member member) : member의 Id와 role이 저장된 토큰을 생성해 반한한다.(토큰의 유효기간은 1시간이다.)

**Comment by wonyongChoi05** [comment]
```java
    /**
     * 최상위 키워드를 만드는 경우, 키워드의 부모 값에 null을 넣어줌
     */
    public Long createKeyword(final Long sessionId, final KeywordCreateRequest request) {
        existSession(sessionId);
        Keyword keywordParent = findKeywordParentOrNull(request.getParentKeywordId());

        Keyword keyword = request.toEntity(sessionId, keywordParent);
        keywordRepository.save(keyword);

        return keyword.getId();
    }
```
createKeyword: 최상위 키워드를 생성한다. 요청으로부터 받은 세션 ID와 키워드 생성 요청 정보를 사용하여 키워드 엔티티를 생성하고 저장한다. 부모 키워드 ID가 제공된 경우 해당 키워드를 찾아 부모로 설정한다.

```java
    @Transactional(readOnly = true)
    public KeywordResponse findKeyword(final Long sessionId, final Long keywordId) {
        existSession(sessionId);
        Keyword keyword = keywordRepository.findById(keywordId)
            .orElseThrow(KeywordNotFoundException::new);

        return KeywordResponse.createResponse(keyword);
    }
```
findKeyword: 특정 키워드 ID를 사용하여 키워드를 조회한다. 요청으로부터 받은 세션 ID와 키워드 ID를 사용하여 키워드를 검색하고, 해당 키워드의 정보를 KeywordResponse 객체로 반환한다.

```java
    @Transactional(readOnly = true)
    public KeywordResponse findKeywordWithAllChild(final Long sessionId, final Long keywordId) {
        existSession(sessionId);
        existKeyword(keywordId);

        Keyword keyword = keywordRepository.findFetchById(keywordId);

        return KeywordResponse.createWithAllChildResponse(keyword);
    }
```
findKeywordWithAllChild: 특정 키워드 ID를 사용하여 해당 키워드와 모든 하위 키워드를 조회한다. 요청으로부터 받은 세션 ID와 키워드 ID를 사용하여 키워드를 검색하고, 해당 키워드와 모든 하위 키워드의 정보를 KeywordResponse 객체로 반환한다.

```java
    @Transactional(readOnly = true)
    public KeywordsResponse findSessionIncludeRootKeywords(final Long sessionId) {
        existSession(sessionId);

        List<Keyword> keywords = keywordRepository.findBySessionIdAndParentIsNull(sessionId);

        return KeywordsResponse.createResponse(keywords);
    }
```
findSessionIncludeRootKeywords: 세션에 속한 최상위 키워드들을 조회한다. 요청으로부터 받은 세션 ID를 사용하여 세션에 속한 최상위 키워드들을 검색하고, 해당 키워드들의 정보를 KeywordsResponse 객체로 반환한다.

```java
    public void updateKeyword(final Long sessionId, final Long keywordId,
                              final KeywordUpdateRequest request) {
        existSession(sessionId); // 세션이 없다면 예외가 발생
        Keyword keyword = keywordRepository.findById(keywordId)
            .orElseThrow(KeywordNotFoundException::new); // keyword 가 없다면 예외를 발생
        Keyword keywordParent = findKeywordParentOrNull(request.getParentKeywordId());

        keyword.update(request.getName(), request.getDescription(), request.getOrder(),
            request.getImportance(), keywordParent);
    }
```
updateKeyword: 특정 키워드를 업데인트한다. 요청으로부터 받은 세션 ID와 키워드 ID를 사용하여 키워드를 검색하고, 요청으로 받은 정보를 사용하여 키워드의 속성을 업데이트한다. 부모 키워드 ID가 제공된 경우 해당 키워드를 찾아 부모로 설정한다.

```java
    public void deleteKeyword(final Long sessionId, final Long keywordId) {
        existSession(sessionId);
        Keyword keyword = keywordRepository.findFetchById(keywordId);

        keywordRepository.delete(keyword);
    }
```
deleteKeyword: 특정 키워드를 삭제한다. 요청으로부터 받은 세션 ID와 키워드 ID를 사용하여 키워드를 검색하고, 키워드를 삭제한다.

```java
    private void existSession(final Long sessionId) {
        boolean exists = sessionRepository.existsById(sessionId);
        if (!exists) {
            throw new SessionNotFoundException();
        }
    }
```
existSession: 주어진 세션 ID로 세션이 존재하는지 확인한다. 세션이 존재하지 않으면 SessionNotFoundException을 발생시킨다.

```java
    private void existKeyword(final Long keywordId) {
        boolean exists = keywordRepository.existsById(keywordId);
        if (!exists) {
            throw new KeywordNotFoundException();
        }
    }
```
existKeyword: 주어진 키워드 ID로 키워드가 존재하는지 확인한다. 키워드가 존재하지 않으면 KeywordNotFoundException을 발생시킨다.

```java
    private Keyword findKeywordParentOrNull(final Long keywordId) {
        if (keywordId == null) {
            return null;
        }
        return keywordRepository.findById(keywordId).orElseThrow(KeywordNotFoundException::new);
    }
```
findKeywordParentOrNull: 주어진 키워드 ID로 키워드를 검색하여 반환한다. 키워드 ID가 null인 경우 null을 반환다. 키워드가 존재하지 않으면 KeywordNotFoundException을 발생시킨다.

**Comment by This2sho** [comment]
# EssayAnswerService

: 퀴즈 답변 관련 서비스 필드로 `essayAnswerRepository`, `quizRepository`, `memberService` 가 존재
내부 기능은 복잡한 로직 없이 CRUD

<img width="690" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/3605a363-371d-499a-bcfb-58ebd2963e20">

관련 화면: 로드맵 → 키워드 클릭

밑에서 부터 기능별 설명

---

### `createEssayAnswer(EssayAnswerRequest essayAnswerRequest, Long memberId)`

<img width="680" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/c3130523-2c78-45f2-8e68-1d0963269eac">

: 퀴즈 질문 선택시 답변 작성 가능

---

### `updateEssayAnswer(Long answerId, String answer, Long memberId)`

<img width="684" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/f62491cc-6674-4684-8397-6155f5d16289">

: 답변 작성후 가보면 있을 거 같지만.. 현재 업데이트 기능은 서비스에 존재하지 않음
(컨트롤러에 해당 기능 존재 x, 테스트에만 사용중)

---

### `deleteEssayAnswer(Long answerId, Long memberId)`

<img width="684" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/3b9484b6-0ba1-4f58-8769-c87174a1ab42">

: 답변 작성후 삭제 가능

---

### `public List<EssayAnswer> findByQuizId(Long quizId)`

<img width="680" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/aaaa3e5a-3322-453e-83fe-a6dd39c1313a">

: 답변 보러가기 선택시 해당 퀴즈와 관련된 Answer 보여줌

---

### `EssayAnswer getById(Long answerId)`

<img width="682" alt="image" src="https://github.com/woowacourse/prolog/assets/48519860/3c16a049-cfa7-4ab9-acae-1f98f058afe5">

: 답변 보러가기에서 특정 답변 선택시 보러갈 수 있음.

**Comment by donghae-kim** [comment]
## LOGIN

### LoginInterceptor

```java
githubLoginService.validateToken(AuthorizationExtractor.extract(request));
```

- AccessToken이 만료기간이 유효하지 않으면 `TokenNotValidException` 발생시킨다.
- 그 외에는 True 반환한다.

### AuthMemberPrincipalArgumentResolver

```java
String credentials = AuthorizationExtractor
            .extract(webRequest.getNativeRequest(HttpServletRequest.class));
```

`HttpServletRequest` 의 `Authorization` 헤더 필드에서 토큰을 추출합니다.

- `Authorization` 헤더 필드에 토큰이 존재한다면 credentials 변수에 할당됩니다.
- 토큰이 존재하지 않는다면 null 이 할당 됩니다.

```java
if (credentials == null || credentials.isEmpty()) {
    memberAuthorityCache.setAuthority(Authority.ANONYMOUS);
    return new LoginMember(Authority.ANONYMOUS);
}
```

`AccessToken` 이 Null or Empty 면 `ANOYMOUS` (비회원 사용자)

- 비회원 사용자를 걸러낸다.
- credentials 가 null 혹은 Empty 면 비회원 사용자이다.
- `memberAuthorityCache`의 Authority 가 `ANONYMOUS` 로 설정된다.

```java
try {
    Long id = Long.parseLong(jwtTokenProvider.extractSubject(credentials));
    memberAuthorityCache.setAuthority(Authority.MEMBER);
    return new LoginMember(id, Authority.MEMBER);
} catch (NumberFormatException e) {
    throw new TokenNotValidException();
}
```

회원인 사용자에 대한 검증을 진행한다.

- 토큰이 있고, 유효한 토큰이면 `MEMBER`
- 토큰이 있는데, 유효하지 않은 토큰이면 `TokenNotValidExcetpion` 발생한다.

### **MemberAuthorityCache**

```java
@Component
@RequestScope
public class MemberAuthorityCache {

    private Authority authority;

    public Authority getAuthority() {
        return authority;
    }

    public void setAuthority(Authority authority) {
        this.authority = authority;
    }
}
```

- 해당 요청을 한 멤버의 권한을 잠시 저장해놓는 객체입니다. (Spring Bean)
- `@ReqeustScope` 가 선언되어 있는데, 이는 스프링 Bean 의 생명주기를 현재 들어온 요청으로 제한하는 것입니다.
- `AuthMemberPrincipalArgumentResolver` 를 통해 `MemberAuthorityCache` 의 Authority를 설정하고, MemberOnly AOP 과정에서 이를 이용해 권한을 검증합니다.

### MemberOnly

```java
@Aspect
@Component
@RequiredArgsConstructor
public class LoginMemberVerifier {

    private final MemberAuthorityCache memberAuthorityCache;

    @Before("@annotation(wooteco.prolog.login.aop.MemberOnly)")
    public void checkLoginMember() {
        final Authority authority = memberAuthorityCache.getAuthority();
        if (!authority.equals(Authority.MEMBER)) {
            throw new MemberNotAllowedException();
        }
    }
}
```

- Spring AOP를 통해 멤버 권한 체크를 진행한다.
- `memberAuthorityCache` 에 저장해둔 권한이 MEMBER가 아니라면 `MemberNotAllowedException` 예외를 발생시킨다.

예를들어

```java
@PutMapping("/{id}")
@MemberOnly
public ResponseEntity<Void> updateLovellog(@AuthMemberPrincipal LoginMember member,
                                           @PathVariable Long id,
                                           @RequestBody LevelLogRequest levelLogRequest) {
    levelLogService.updateLevelLog(member.getId(), id, levelLogRequest);
    return ResponseEntity.noContent().build();
}
```

- `@MemberOnly` 어노테이션이 존재하는 메소드의 실행 전에 위의 `checkLoginMember`가 실행된다.

원래 있던 위키에도 잘 정리가 되어있습니다!
[WIKI](https://github.com/woowacourse/prolog/wiki/%EC%9D%B8%EA%B0%80-%EA%B3%BC%EC%A0%95)
